airflow/utils/log/secrets_masker.py

```diff
@@ -0,0 +1,221 @@
+    """Get comma-separated sensitive Variable Fields from airflow.cfg."""
+    from airflow.configuration import conf
+
+    sensitive_fields = DEFAULT_SENSITIVE_FIELDS.copy()
+    sensitive_variable_fields = conf.get('core', 'sensitive_var_conn_names')
+    if sensitive_variable_fields:
+        sensitive_fields |= frozenset({field.strip() for field in sensitive_variable_fields.split(',')})
+    return sensitive_fields
+
+
+def should_hide_value_for_key(name):
+    """Should the value for this given name (Variable name, or key in conn.extra_dejson) be hidden"""
+    from airflow import settings
+
+    if name and settings.HIDE_SENSITIVE_VAR_CONN_FIELDS:
+        name = name.strip().lower()
+        return any(s in name for s in get_sensitive_variables_fields())
+    return False
+
+
+def mask_secret(secret: Union[str, dict, Iterable], name: str = None) -> None:
+    """
+    Mask a secret from appearing in the task logs.
+
+    If ``name`` is provided, then it will only be masked if the name matches
+    one of the configured "sensitive" names.
+
+    If ``secret`` is a dict or a iterable (excluding str) then it will be
+    recursively walked and keys with sensitive names will be hidden.
+    """
+    # Delay import
+    from airflow import settings
+
+    # Filtering all log messages is not a free process, so we only do it when
+    # running tasks
+    if not settings.MASK_SECRETS_IN_LOGS or not secret:
+        return
+
+    _secrets_masker().add_mask(secret, name)
+
+
+def redact(value: "RedactableItem", name: str = None) -> "RedactableItem":
+    """Redact any secrets found in ``value``."""
+    return _secrets_masker().redact(value, name)
+
+
+@cache
+def _secrets_masker() -> "SecretsMasker":
+
+    for flt in logging.getLogger('airflow.task').filters:
+        if isinstance(flt, SecretsMasker):
+            return flt
+    raise RuntimeError("No SecretsMasker found!")
+
+
+class SecretsMasker(logging.Filter):
+    """Redact secrets from logs"""
+
+    replacer: Optional["RePatternType"] = None
+    patterns: Set[str]
+
+    ALREADY_FILTERED_FLAG = "__SecretsMasker_filtered"
+
+    def __init__(self):
+        super().__init__()
+        self.patterns = set()
+
+    @cached_property
+    def _record_attrs_to_ignore(self) -> Iterable[str]:
+        # Doing log.info(..., extra={'foo': 2}) sets extra properties on
+        # record, i.e. record.foo. And we need to filter those too. Fun
+        #
+        # Create a record, and look at what attributes are on it, and ignore
+        # all the default ones!
+
+        record = logging.getLogRecordFactory()(
+            # name, level, pathname, lineno, msg, args, exc_info, func=None, sinfo=None,
+            "x",
+            logging.INFO,
+            __file__,
+            1,
+            "",
+            tuple(),
+            exc_info=None,
+            func="funcname",
+        )
+        return frozenset(record.__dict__.keys()) - frozenset(('msg', 'args'))
```

Comment:
Man, I really hate how this looks, but it seems there really isn’t a better way. Keys in `extra` are lost in `Logger` and unless we overwrite `Logger.manager` (which would *greatly* complicate things) there isn’t a way to get the information.

As a nitpick, it’s probably easier to read (and maybe slightly more performant) to do

```python
frozenset(record.__dict__).difference(("msg", "args"))
```

(But maybe it’s just I’ve never properly learned to read set operators.)

Reply:
Yeah, I was not happy about having to do this, but somewhere in stdlib logging it does `setattr(record, k, val)` with the extra kwargs.

Yes, sure can change to `difference`

Reply:
https://github.com/python/cpython/blob/9ee8448243e776d2a07a9868e9795bbb2c828f9c/Lib/logging/__init__.py#L1588 is where it's done in std python lib :(

Reply:
It could have *at least* set `rv.extra = extra` or something, but somehow it feels the need to set the attributes and then discard the context entirely. Sad. (The `logging` module in general leaves a lot to be desired, really.)